### PR TITLE
[SPARK-20946][SQL] Simplify the config setting logic in SparkSession.getOrCreate

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
# What changes were proposed in this pull request?

The current conf setting logic is a little complex and has duplication, this PR simplifies it.
This is a second try of the previous PR, #18172 , of @cloud-fan . Only one line is updated. All credit should go to @cloud-fan .

This also closes #18501 .

# How was this patch tested?

Pass the Jenkins with existing tests.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan and @gatorsmile .
Could you review this PR?

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 1]
```diff
@@ -904,22 +906,14 @@ object SparkSession {
 
         // No active nor global default session. Create a new one.
         val sparkContext = userSuppliedContext.getOrElse {
-          // set app name if not given
-          val randomAppName = java.util.UUID.randomUUID().toString
           val sparkConf = new SparkConf()
           options.foreach { case (k, v) => sparkConf.set(k, v) }
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: This line is changed. `options` are copied into new SparkConf.

**Comment by dongjoon-hyun** [comment]
The failure seems to be irrelevant.
```
org.apache.spark.sql.kafka010.KafkaSourceSuite.subscribing topic by name from latest offsets (failOnDataLoss: true)
```

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by cloud-fan** [comment]
LGTM, thanks for picking it up!

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 2]
```diff
@@ -904,22 +906,14 @@ object SparkSession {
...
-          if (!sc.conf.contains("spark.app.name")) {
-            sc.conf.setAppName(randomAppName)
-          }
-          sc
+          // set a random app name if not given.
+          sparkConf.setAppName(options.getOrElse("spark.app.name",
+            java.util.UUID.randomUUID().toString))
+          SparkContext.getOrCreate(sparkConf)
         }
+        options.foreach { case (k, v) => sparkContext.conf.set(k, v) }
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: The only concern is `sparkContext` could be shared by multiple sessions.

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: Then, how do you think about #18501  using `initialSessionOptions`? That is used only once.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: We already did this in https://github.com/apache/spark/pull/18512/files#diff-d91c284798f1c98bf03a31855e26d71cL917 .

The only difference is about `userSuppliedContext`, which is given by a [private method](https://github.com/apache/spark/pull/18512/files#diff-d91c284798f1c98bf03a31855e26d71cR767), so we should be able to check if it's safe or not.

**Comment by dongjoon-hyun** [comment]
@gatorsmile and @cloud-fan .
Please let me know if I need to do something more in this PR. :)

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]:
```
    sparkContext.stop()
    val conf = new SparkConf().setAppName("test").setMaster("local").set("key1", "value1")
    val sparkContext2 = new SparkContext(conf)
    val session =
      SparkSession.builder().config("key2", "value2").sparkContext(sparkContext2).getOrCreate()
    assert(session.conf.get("key1") == "value1")
    assert(session.conf.get("key2") == "value2")
    assert(session.sparkContext.conf.get("key1") == "value1")
    assert(session.sparkContext.conf.get("key2").isEmpty) // <-- This line will fail after the changes of this PR.
    assert(session.sparkContext.conf.get("spark.app.name") == "test")
    session.stop()
```

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: The existing behavior is weird. See https://github.com/apache/spark/pull/18517/

**Pull request closed without merging**